File: package/rust/src/langtons_ant.rs
use std::collections::HashSet;

#[derive(Clone, Copy, PartialEq)]
enum Direction {
    North = 0,
    East = 1,
    South = 2,
    West = 3,
}

impl Direction {
    fn turn_right(self) -> Direction {
        match self {
            Direction::North => Direction::East,
            Direction::East => Direction::South,
            Direction::South => Direction::West,
            Direction::West => Direction::North,
        }
    }

    fn turn_left(self) -> Direction {
        match self {
            Direction::North => Direction::West,
            Direction::East => Direction::North,
            Direction::South => Direction::East,
            Direction::West => Direction::South,
        }
    }

    fn get_delta(self) -> (i32, i32) {
        match self {
            Direction::North => (0, 1),
            Direction::East => (1, 0),
            Direction::South => (0, -1),
            Direction::West => (-1, 0),
        }
    }
}

pub fn langtons_ant(steps: i32) -> Vec<(i32, i32)> {
    let mut black_cells = HashSet::new();
    let mut ant_x = 0;
    let mut ant_y = 0;
    let mut direction = Direction::North;

    for _ in 0..steps {
        if black_cells.contains(&(ant_x, ant_y)) {
            black_cells.remove(&(ant_x, ant_y));
            direction = direction.turn_right();
        } else {
            black_cells.insert((ant_x, ant_y));
            direction = direction.turn_left();
        }

        let (dx, dy) = direction.get_delta();
        ant_x += dx;
        ant_y += dy;
    }

    let mut result: Vec<(i32, i32)> = black_cells.into_iter().collect();
    result.sort();
    result
}

File: package/php/src/LangtonsAnt.php
<?php

declare(strict_types=1);

namespace ProgramKnock;

class LangtonsAnt
{
    /**
     * ラングトンのアリのシミュレーション
     *
     * @param int $steps アリが移動するステップ数
     * @return array<array{0: int, 1: int}> 黒いマスの座標のリスト（ソート済み）
     */
    public static function simulate(int $steps): array
    {
        // 方向: 0=上(北), 1=右(東), 2=下(南), 3=左(西)
        $directions = [
            [0, 1],   // 上: y+1
            [1, 0],   // 右: x+1
            [0, -1],  // 下: y-1
            [-1, 0]   // 左: x-1
        ];

        $antX = 0;
        $antY = 0;
        $antDirection = 0; // 初期方向は上（北）

        // 黒いマスを記録するセット（連想配列で実装）
        $blackCells = [];

        for ($i = 0; $i < $steps; $i++) {
            $key = "$antX,$antY";

            if (isset($blackCells[$key])) {
                // 黒いマスの場合
                unset($blackCells[$key]); // 白に変更
                $antDirection = ($antDirection + 3) % 4; // 左に90度回転（-1を+3で表現）
            } else {
                // 白いマスの場合
                $blackCells[$key] = true; // 黒に変更
                $antDirection = ($antDirection + 1) % 4; // 右に90度回転
            }

            // 1マス前進
            $antX += $directions[$antDirection][0];
            $antY += $directions[$antDirection][1];
        }

        // 黒いマスの座標をリストに変換
        $result = [];
        foreach (array_keys($blackCells) as $key) {
            list($x, $y) = explode(',', $key);
            $result[] = [(int)$x, (int)$y];
        }

        // (x, y) の昇順でソート
        usort($result, function($a, $b) {
            if ($a[0] !== $b[0]) {
                return $a[0] - $b[0];
            }
            return $a[1] - $b[1];
        });

        return $result;
    }
}
